Add XPath and JSON Pointer path formatters

david-waltermire opened the pull request:
## Summary

This PR adds new path formatters and format-based selection to the Metaschema framework:

- **XPathFormatter** - Produces namespace-qualified XPath 3.1 paths using EQName format
- **JsonPointerFormatter** - Produces RFC 6901 compliant JSON Pointer paths
- **Format-based selection** - Automatically uses appropriate path format based on document type

Addresses metaschema-framework/oscal-cli#32

### Key Features

**Path Formatters:**
- XPath with EQName namespace qualification (`Q{namespace}element[position]`)
- Proper handling of XML grouping (GROUPED/UNGROUPED)
- JSON Pointer with proper handling of all JSON grouping behaviors (KEYED/LIST/SINGLETON_OR_LIST/NONE)
- Support for JSON key flag values in KEYED collections
- RFC 6901 escaping (`~` → `~0`, `/` → `~1`)

**Format-Based Selection:**
- JSON/YAML documents automatically use JSON Pointer paths
- XML documents automatically use XPath paths
- New `--path-format` CLI option to override: `auto`, `metapath`, `xpath`, `jsonpointer`

## Changes

### Core Module
- Add `XPathFormatter` - XPath 3.1 paths with EQName-qualified names
- Add `JsonPointerFormatter` - RFC 6901 compliant JSON Pointer paths
- Add `PathFormatSelection` enum for path format options
- Add `XPATH_PATH_FORMATTER` and `JSON_POINTER_PATH_FORMATTER` constants to `IPathFormatter`
- Remove obsolete `JsonPathFormatter` placeholder code
- Update package-info.java documentation

### Databind Module
- Add `Format.getPathFormatter()` - returns appropriate formatter for document format
- Add `Format.resolvePathFormatter(PathFormatSelection, Format)` - resolves formatter based on selection

### CLI Module
- Add `--path-format` option to validation commands
- Update `LoggingValidationHandler` to use configurable path formatter

## Test plan

- [x] 26 unit tests for XPathFormatter covering all node types and edge cases
- [x] 21 unit tests for JsonPointerFormatter covering all grouping behaviors and RFC 6901 escaping
- [x] 10 unit tests for Format path formatter resolution
- [x] Full test suite passes
- [x] CI build passes (`mvn install -PCI -Prelease`)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added XPath 3.1 EQName-based path formatter for XML paths
  * Added RFC 6901 JSON Pointer formatter for JSON/YAML paths
  * Added `--path-format` CLI option for selecting path output format (auto, metapath, xpath, jsonpointer)
  * Validation output now displays paths in selected format

* **Removals**
  * Removed legacy JsonPathFormatter

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

david-waltermire committed:
Add PRD for XPath and JSON Pointer path formatters

david-waltermire committed:
Add XPath and JSON Pointer path formatters

Implement two new IPathFormatter implementations:

- XPathFormatter: Produces XPath 3.1 paths with EQName format
  (e.g., /Q{http://example.com}catalog/Q{http://example.com}control[1])
  Handles GROUPED/UNGROUPED XmlGroupAsBehavior for wrapper elements

- JsonPointerFormatter: Produces RFC 6901 compliant JSON Pointer paths
  (e.g., /catalog/controls/0/id)
  Handles all JsonGroupAsBehavior modes (NONE, LIST, KEYED, SINGLETON_OR_LIST)
  Implements proper RFC 6901 escaping (~ -> ~0, / -> ~1)

Add constants to IPathFormatter:
- XPATH_PATH_FORMATTER
- JSON_POINTER_PATH_FORMATTER

Remove obsolete JsonPathFormatter (was commented out placeholder code).

Update package-info.java with documentation for new formatters.

david-waltermire committed:
Update PRD with Phase 4: Format-based selection and CLI integration

Add design for:
- PathFormatSelection enum (AUTO, METAPATH, XPATH, JSON_POINTER)
- IPathFormatter.resolveFormatter() for format-based selection
- CLI --path-format option for explicit override
- Default behavior uses AUTO (format-based selection)

Mark Phases 1-3 as completed, add Phase 4 implementation tasks and test cases.

david-waltermire committed:
Add format-based path formatter selection and CLI option

Implement automatic path format selection based on document type:
- JSON/YAML documents use JSON Pointer paths by default
- XML documents use XPath paths by default

Add --path-format CLI option to override the default:
- auto (default): select based on document format
- metapath: force Metapath format
- xpath: force XPath 3.1 EQName format
- jsonpointer: force RFC 6901 JSON Pointer format

Changes:
- Add PathFormatSelection enum in core module
- Add Format.getPathFormatter() and Format.resolvePathFormatter() in databind
- Add --path-format option to AbstractValidateContentCommand
- Update LoggingValidationHandler to accept configurable path formatter
- Add unit tests for Format path formatter resolution

david-waltermire committed:
Address PR review comments

- Fix PRD markdown code fence language identifiers (lines 105, 170)
- Fix typo: METAPATH_PATH_FORMATER -> METAPATH_PATH_FORMATTER in PRD
- Clarify PRD tasks for countSiblings() and getJsonKeyValue() helpers
- Add warning logging when getJsonKeyValue() falls back to numeric index

aj-stein reviewed (dismissed):
Very useful, I look forward to using this feature.

david-waltermire committed:
Integrate --path-format option with shell completion

Add .type(PathFormatSelection.class) to PATH_FORMAT_OPTION and
register PathFormatSelection enum in CompletionTypeRegistry for
shell completion support from PR 551.

david-waltermire commented on PRDs/20251215-path-formatters/plan.md:
````diff
@@ -0,0 +1,493 @@
+AbstractConstraintValidationHandler.setPathFormatter()
+    ↓ used by
+Validation result messages
+```
````
done

david-waltermire commented on core/src/main/java/gov/nist/secauto/metaschema/core/metapath/format/JsonPointerFormatter.java:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  @Override
+  @NonNull
+  public String formatDocument(IDocumentNodeItem document) {
+    // Returns empty string to produce leading "/" via join in format method
+    return "";
+  }
+
+  @Override
+  @NonNull
+  public String formatRootAssembly(IRootAssemblyNodeItem root) {
+    String jsonName = root.getDefinition().getJsonName();
+    return escapeJsonPointer(jsonName);
+  }
+
+  @Override
+  @NonNull
+  public String formatAssembly(IAssemblyNodeItem assembly) {
+    return formatModelItem(assembly);
+  }
+
+  @Override
+  @NonNull
+  public String formatAssembly(IAssemblyInstanceGroupedNodeItem assembly) {
+    return formatModelItem(assembly);
+  }
+
+  @Override
+  @NonNull
+  public String formatField(IFieldNodeItem field) {
+    return formatModelItem(field);
+  }
+
+  @Override
+  @NonNull
+  public String formatFlag(IFlagNodeItem flag) {
+    // JSON Pointer does not use @ prefix for attributes
+    return escapeJsonPointer(flag.getQName().getLocalName());
+  }
```
Override methods inherit docs.
